src/plugins/intel_cpu/src/nodes/kernels/registers_pool.hpp

```diff
@@ -0,0 +1,184 @@
+     * It could be created by using constructor with RegistersPool as an argument, like the next:
+     *      const RegistersPool::Reg<Xbyak::Reg64> reg {regPool};
+     * The destructor will return the register to the pool.
+     */
+    template<class TReg>
+    class Reg {
+        friend class RegistersPool;
+    public:
+        Reg(RegistersPool& regPool);
+        ~Reg() { if (onDestruct) { onDestruct(); } }
+        Reg& operator=(Reg&& other)  noexcept {
+            reg = other.reg;
+            onDestruct = std::move(other.onDestruct);
+            return *this;
+        }
+        Reg(Reg&& other)  noexcept : reg(other.reg), onDestruct(std::move(other.onDestruct)) {}
+        operator TReg&() { assert(isValid()); return reg; }
+        operator const TReg&() const { assert(isValid()); return reg; }
+        operator Xbyak::RegExp() const { assert(isValid()); return reg; }
+        int getIdx() const { assert(isValid()); return reg.getIdx(); }
+        friend Xbyak::RegExp operator+(const Reg& lhs, const Xbyak::RegExp& rhs) {
+            assert(lhs.isValid());
+            return lhs.operator Xbyak::RegExp() + rhs;
+        }
+        bool isValid() const { return onDestruct.operator bool(); }
+
+    private:
+        Reg(const TReg& reg, const std::function<void()>& onDestruct)
+                : reg(reg), onDestruct(onDestruct) {}
+
+    private:
+        TReg reg;
+        std::function<void()> onDestruct;
+    };
+
+    virtual ~RegistersPool() = default;
+
+protected:
+    struct PhysicalSet {
+        std::unordered_set<int> unusedIndexes;
+        int maxIndexValue;
+
+        PhysicalSet(int maxIndexValue)
+                : maxIndexValue(maxIndexValue) {
+            for (int i = 0; i <= maxIndexValue; ++i)
+                unusedIndexes.insert(i);
+        }
+
+        void setAsUsed(int regIdx) {
+            assert(regIdx <= maxIndexValue && regIdx >= 0);
+            auto it = unusedIndexes.find(regIdx);
+            assert(it != unusedIndexes.end());
+            unusedIndexes.erase(it);
+        }
+
+        void setAsUnused(int regIdx) {
+            assert(regIdx <= maxIndexValue && regIdx >= 0);
+            unusedIndexes.insert(regIdx);
+        }
+
+        int getUnused() {
+            assert(!unusedIndexes.empty());
+            return *unusedIndexes.begin();
+        }
+
+        void exclude(Xbyak::Reg reg) {
+            unusedIndexes.erase(reg.getIdx());
+        }
+    };
+
+    RegistersPool(std::initializer_list<Xbyak::Reg> regsToExclude, int simdRegistersNumber)
+            : simdSet(simdRegistersNumber-1) {
+        using namespace Xbyak;
+        for (auto& reg : regsToExclude) {
+            if (reg.isYMM() || reg.isXMM() || reg.isZMM()) {
+                simdSet.exclude(reg);
+            } else if (reg.isOPMASK()) {
+                excludeOpmask(reg);
+            } else if (reg.isREG()) {
+                generalSet.exclude(reg);
+            }
+        }
+    }
+
+    template<class T>
```

Comment:
Please, use `typename` vs `class` keyword as more common approach in code base

Reply:
Done